### PR TITLE
feat: Adds cypress based fakeAPI mocking plus cucumber usage example

### PR DESCRIPTION
**Pull request by johncowen**
This PR uses Cypress as a browser based test runner to run our more e2e-like tests. Additionally it does this by using cucumber as a thin layer over test assertions etc. Before explaining further, I just wanted to point out that when I say 'e2e', I mean 'frontend e2e' i.e. these tests should not use any backend infrastructure for testing (at least for the moment) These are 'e2e' tests, but without testing the things that we don't own.

We have a longish running desire to run our e2e-like tests in a browser rather than using Node/CLI via Jest, this PR is an approach to that task.

I've used the existing CLI based e2e-like tests from DataPlaneListView.spec (https://github.com/kumahq/kuma-gui/blob/1c7311e7d093397b853ce7f50dd224f7daf13493/src/app/data-planes/views/DataPlaneListView.spec.ts) as a starting point for browser based testing, and you'll see that the new tests here replicate the DataplaneListView tests that we are currently using a CLI to test. I've immediately seen one benefit in using this approach as it prevents things like:

https://github.com/kumahq/kuma-gui/blob/1c7311e7d093397b853ce7f50dd224f7daf13493/src/app/data-planes/views/DataPlaneListView.spec.ts#L292

where we've told the tests to `expect` without performing an assertion, so the above line currently doesn't assert anything in the places that we use it. In cucumber you'd get an error if you did something similar saying the step doesn't exist.

Another benefit seems to be non-clashing jest/cypress types.

There is work still todo here which I would like to do as a relatively soon follow up task[s], but it's currently at a point where we move all of our e2e tests forwards without more noodling with code (there are typescript types to sort out and probably a bit of duplication to remove). I've kept all the code that could potentially be noodled with in the one place in `@/test-support`, and as we are using cucumber for syntax we can noodle all we like with the underlying code (when the time is right) and it doesn't effect any of our potentially 1000s of tests.

Briefly:

1. The main bit of work was getting our dynamic mocks to work within Cypress, but this time using `cy.intercept` rather than MSW. It doesn't look like Cypress can perform the variable substitution from URLs for you (i.e. `/meshes/:mesh/` > `{mesh: "default"}`, so I've used URLPattern to do that since I'm also using that for something else/elsewhere. This would also be useful on the MSW side of things.
2. We currently use MSW types for a Request data object (i.e. `params` and `query` etc). For the moment I've just `@ts-ignore`d the places where the types originating from Cypress don't match the MSW types we currently use. These will need making more generic/less MSW-coupled when we come back to noodle with this. (edit: see comment I make below, whilst waiting on review I've noodled a bit and I think I've sorted this)
3. We will need to use our `@/services/development` file so we can re-use all the mock file setup (aswell as the ability to label these for them to automatically picked up) similar to what we do during development/jest testing. Currently Cypress gives a tonne of log errors related to loading `.vue` files. There is a task here https://github.com/kumahq/kuma-gui/issues/798 to split development out so it doesn't also load `production` so that might help here. Although it might also be better to configure Cypress to load these if that's possible. I'm unsure about this right now/still thinking.
4. `env` will also need to come from the main application (as it does when using MSW)
5. Probably more comments... (I'll edit in here if/when I think of them)

All in all I wanted to get this up now for 👀 rather than noodle more.

This is not set up in CI yet and shouldn't be until we are a bit further ahead with it, so for the moment you should run `yarn dev:real-api` (we don't want MSW mocking) on port `8080` and run `npx cypress open` in order to see this working (you might have to do the ol' `cypress install` thing previous to that also)

As always, any questions around getting it running or anything else please ask. We also have a GUI sync up tomorrow so it might be a good opportunity to discuss this further then.

### CODE REVIEW

File: src/test-support/mocks/src/dataplanes+insights.ts [hunk 1]
```diff
@@ -1,10 +1,9 @@
 import type { EndpointDependencies, MockResponder } from '@/test-support'
 export default ({ fake, pager, env }: EndpointDependencies): MockResponder => (req) => {
-  const params = req.params
   const { offset, total, next, pageTotal } = pager(
     env('KUMA_DATAPLANE_COUNT', `${fake.datatype.number({ min: 1, max: 1000 })}`),
     req,
-    `/meshes/${params.mesh}/dataplanes+insights`,
+    '/dataplanes+insights',
```
Comment (johncowen) [inline src/test-support/mocks/src/dataplanes+insights.ts:6]: I spotted this was wrong so fixed this up whilst I was here, would be nice for `Pager` to automatically figure these out in the future.

Comment (johncowen) [reply src/test-support/mocks/src/dataplanes+insights.ts:6]: I think there's a couple. One is at root based and one is mesh based:

https://github.com/kumahq/kuma-gui/blob/1c7311e7d093397b853ce7f50dd224f7daf13493/src/services/kuma-api/KumaApi.ts#L134-L140

This one is the root one. Not totally sure if we still use the root based one in the GUI. 👀 

At first glance it looks like we don't so we could completely remove the root one (and any other similar root ones).

Comment (lahabana) [reply src/test-support/mocks/src/dataplanes+insights.ts:6]: Ok feel free to open an issue to figure this out later and ignore this comment :)

**Comment by johncowen** [comment]
After PRing this I tried some aforementioned noodling to remove some duplication and remove the `@ts-ignore`.

Completely informal work in progress right now, but here's a link to that https://github.com/johncowen/kuma-gui/compare/feature/cypress...johncowen:kuma-gui:chore/cypress-refact I'm likely to PR this separately soon after. Thought I'd mention here and give a changeset to view even though I can't nest a PR here. I'll try and describe more in the desc for it once I PR it.

**Comment by johncowen** [comment]
I think the other thing I wanted to ask here was: What would next steps be now that we are in a place where we can standup a production build and run e2e tests on it in a browser using mocks?

Options (there may be more than this)

1. CI integration
2. Convert our current CLI based tests to browser based
3. Start creating net new tests (and if so are there any places in particular where we would like to start adding i.e. are there places that are currently lighter in tests than other areas?)

For the CI integration, this will potentially involve standing up/serving a build in CI. We can't use the preview site as they are MSW-mocked by default  (unless we can base the MSW feature flag on a runtime value, say a based on a cookie or something)

**Comment by johncowen** [comment]
Before I forget there's at least one TODO to do here before we merge:

1. Take a look at moving the `.feature` files into the `/cypress` folder instead of being colocated along with the Route/View Component file. We only have one feature file at the minute so shouldn't be an issue

Just to note that ^ shouldn't block review but should be looked at before merge.

**Comment by kleinfreund** [comment]
FYI, you will need to rebase/merge with master in order to run the tests using the updated GitHub Actions workflow since a job in it was renamed.

This is necessary because branch protection rules for which status checks are required are based on job names and so currently, this PR is waiting for a "Tests" job to have succeeded when it was previously called "Tests & build".

**Review by kleinfreund — APPROVED** [review]

**Comment by johncowen** [comment]
Just wanted to mention before merging this, that there are 2 follow up PRs for this mentioned above - the last one should be ready to go, the previous might be a little out of date and may need a bit of massaging before PRing.

Apart from that I should be able to work on re-using our `development` service config now `development` doesn't import `production`. I can also start looking at bringing this all into CI.

**Pull request merged**